Improve LDAP setup workflow

bshepherdson opened the pull request:
Goal is to separate configuration from enabling.
If a new configuration validates successfully, LDAP is enabled. After
that it can be turned off without damaging the configuration.

The frontend changes are a hack to test this somewhat and can't be
checked in.

Backend changes for #19428, umbrella bug is #16225.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/21888)
<!-- Reviewable:end -->

bshepherdson committed:
First draft of slicker LDAP authentication

Goal is to separate configuration from enabling.
If a new configuration validates successfully, LDAP is enabled. After
that it can be turned off without damaging the configuration.

The frontend changes are a hack to test this somewhat and can't be
checked in.

noahmoss commented:
Nice, I think these changes make sense. Basically you'll only ever be able to save LDAP details that can be validated, regardless of whether it's actively enabled.

Are there any downsides to making this change even with the current UI?

bshepherdson commented:
Yes, actually. If my reckoning is right this change with the existing frontend makes it impossible to configure LDAP. (Can't configure without the enable switch being set; can't enable without a valid configuration.)

The correct FE flow is not hard but requires nontrivial changes. The toggle should be on the list of auth methods, and automatically get set at least some of the time. Probably we should combine FE changes in this branch and land them together.

bshepherdson committed:
Fix and expand tests for new LDAP workflow

bshepherdson committed:
Add /ldap/enable endpoint, and better tests for the logic

**NB:** I'm dropping auto-enabling from the backend!
That's not an API-level concern, and is a weird special case.
The frontend can follow a successful config save with a call to /enable
in cases where that makes sense in the UI (like first-time configuration).

bshepherdson commented:
@tsmacdonald Yes, safe I think. This was a very early bug I explored in my first week or two at Metabase, and when it ran into unexpected difficulties, I was pulled to work on something else. :grin:

It was fixed properly in the fall of 2022 for 45.
